lib/mondrian_redis_segment_cache/cache.rb

```diff
@@ -261,55 +151,77 @@ def has_expiry?
-            mondrian_cache.publish_created_to_listeners(message)
-          end
-        end
+    def publish_created_to_listeners(message)
+      listeners.each do |listener|
+        publish_created_to_listener(message, listener)
       end
     end
 
-    def register_deleted_listener
-      @deleted_listener_thread = Thread.new(deleted_listener_connection, self) do |deleted_redis_connection, mondrian_cache|
-        deleted_redis_connection.subscribe(mondrian_cache.deleted_event_key) do |on|
-          on.message do |channel, message|
-            mondrian_cache.publish_deleted_to_listeners(message)
-          end
-        end
+    def publish_deleted_to_listener(message, listener)
+      segment_header = segment_header_from_base64(message)
+
+      if segment_header
+        deleted_event = ::MondrianRedisSegmentCache::DeletedEvent.new(segment_header)
+        listener.handle(deleted_event)
       end
     end
 
-    def register_expired_listener
-      @expired_listener_thread = Thread.new(expired_listener_connection, self) do |expired_redis_connection, mondrian_cache|
-        expired_redis_connection.subscribe(mondrian_cache.expired_event_key) do |on|
-          on.message do |channel, message|
-            mondrian_cache.publish_expired_to_listeners(message)
-          end
-        end
+    def publish_deleted_to_listeners(message)
+      listeners.each do |listener|
+        publish_deleted_to_listener(message, listener)
       end
     end
 
-    def register_evicted_listener
-      @evicted_listener_thread = Thread.new(evicted_listener_connection, self) do |evicted_redis_connection, mondrian_cache|
-        evicted_redis_connection.subscribe(mondrian_cache.evicted_event_key) do |on|
-          on.message do |channel, message|
-            mondrian_cache.publish_evicted_to_listeners(message)
-          end
+    def reconcile_local_set_with_redis
+      remote_set = Set.new
+
+      mondrian_redis.with do |connection|
+        connection.sscan_each(SEGMENT_HEADERS_SET_KEY) do |segment_header_base64|
+          remote_set << segment_header_base64
         end
       end
+
+      remote_added_keys = remote_set - local_cache_set
+      remote_removed_keys = local_cache_set - remote_set
+      @local_cache_set = remote_set
+
+      remote_added_keys.each do |remote_added_key|
+        publish_created_to_listeners(remote_added_key)
+      end
+
+      remote_removed_keys.each do |remote_removed_key|
+        publish_deleted_to_listeners(remote_removed_key)
+      end
     end
 
+    def reconcile_set_and_keys
+      headers = []
+      mondrian_redis.with do |connection|
+        connection.sscan_each(SEGMENT_HEADERS_SET_KEY) do |segment_header_base64|
+          headers << segment_header_base64
+        end
+      end
+
+      headers.each do |header|
+        # Spin through Header Set and remove any keys that are not in redis at all (they may have been deleted while offline)
+        next if mondrian_redis.with do |connection|
```

Comment:
it was done that way initially but I changed it because I added the `with` .... I should probably update to be a little better as I didn't want 2 connection checkouts in the same block